In [157]:
import numpy as np
from sklearn import model_selection
from sklearn import preprocessing
import pandas as pd
import matplotlib.pyplot as plt


Data analysis and Preprocessing

In [158]:
data = pd.read_csv("diabetes.csv")
data.pop('Sex')
data.pop('Age')
data.pop("Income")
data.pop('Education')

data

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                  0.0     1.0       0.0        1.0  26.0     0.0     0.0   
1                  0.0     1.0       1.0        1.0  26.0     1.0     1.0   
2                  0.0     0.0       0.0        1.0  26.0     0.0     0.0   
3                  0.0     1.0       1.0        1.0  28.0     1.0     0.0   
4                  0.0     0.0       0.0        1.0  29.0     1.0     0.0   
...                ...     ...       ...        ...   ...     ...     ...   
70687              1.0     0.0       1.0        1.0  37.0     0.0     0.0   
70688              1.0     0.0       1.0        1.0  29.0     1.0     0.0   
70689              1.0     1.0       1.0        1.0  25.0     0.0     0.0   
70690              1.0     1.0       1.0        1.0  18.0     0.0     0.0   
70691              1.0     1.0       1.0        1.0  25.0     0.0     0.0   

       HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump  \
0                       0.0           1.0     0.0      1.0                0.0   
1                       0.0           0.0     1.0      0.0                0.0   
2                       0.0           1.0     1.0      1.0                0.0   
3                       0.0           1.0     1.0      1.0                0.0   
4                       0.0           1.0     1.0      1.0                0.0   
...                     ...           ...     ...      ...                ...   
70687                   0.0           0.0     0.0      1.0                0.0   
70688                   1.0           0.0     1.0      1.0                0.0   
70689                   1.0           0.0     1.0      0.0                0.0   
70690                   0.0           0.0     0.0      0.0                0.0   
70691                   1.0           1.0     1.0      0.0                0.0   

       AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  
0                1.0          0.0      3.0       5.0      30.0       0.0  
1                1.0          0.0      3.0       0.0       0.0       0.0  
2                1.0          0.0      1.0       0.0      10.0       0.0  
3                1.0          0.0      3.0       0.0       3.0       0.0  
4                1.0          0.0      2.0       0.0       0.0       0.0  
...              ...          ...      ...       ...       ...       ...  
70687            1.0          0.0      4.0       0.0       0.0       0.0  
70688            1.0          0.0      2.0       0.0       0.0       1.0  
70689            1.0          0.0      5.0      15.0       0.0       1.0  
70690            1.0          0.0      4.0       0.0       0.0       1.0  
70691            1.0          0.0      2.0       0.0       0.0       0.0  

[70692 rows x 18 columns]

In [159]:
data.describe()

Diabetes_binary        HighBP      HighChol     CholCheck  \
count     70692.000000  70692.000000  70692.000000  70692.000000   
mean          0.500000      0.563458      0.525703      0.975259   
std           0.500004      0.495960      0.499342      0.155336   
min           0.000000      0.000000      0.000000      0.000000   
25%           0.000000      0.000000      0.000000      1.000000   
50%           0.500000      1.000000      1.000000      1.000000   
75%           1.000000      1.000000      1.000000      1.000000   
max           1.000000      1.000000      1.000000      1.000000   

                BMI        Smoker        Stroke  HeartDiseaseorAttack  \
count  70692.000000  70692.000000  70692.000000          70692.000000   
mean      29.856985      0.475273      0.062171              0.147810   
std        7.113954      0.499392      0.241468              0.354914   
min       12.000000      0.000000      0.000000              0.000000   
25%       25.000000      0.000000      0.000000              0.000000   
50%       29.000000      0.000000      0.000000              0.000000   
75%       33.000000      1.000000      0.000000              0.000000   
max       98.000000      1.000000      1.000000              1.000000   

       PhysActivity        Fruits       Veggies  HvyAlcoholConsump  \
count  70692.000000  70692.000000  70692.000000       70692.000000   
mean       0.703036      0.611795      0.788774           0.042721   
std        0.456924      0.487345      0.408181           0.202228   
min        0.000000      0.000000      0.000000           0.000000   
25%        0.000000      0.000000      1.000000           0.000000   
50%        1.000000      1.000000      1.000000           0.000000   
75%        1.000000      1.000000      1.000000           0.000000   
max        1.000000      1.000000      1.000000           1.000000   

       AnyHealthcare   NoDocbcCost       GenHlth      MentHlth      PhysHlth  \
count   70692.000000  70692.000000  70692.000000  70692.000000  70692.000000   
mean        0.954960      0.093914      2.837082      3.752037      5.810417   
std         0.207394      0.291712      1.113565      8.155627     10.062261   
min         0.000000      0.000000      1.000000      0.000000      0.000000   
25%         1.000000      0.000000      2.000000      0.000000      0.000000   
50%         1.000000      0.000000      3.000000      0.000000      0.000000   
75%         1.000000      0.000000      4.000000      2.000000      6.000000   
max         1.000000      1.000000      5.000000     30.000000     30.000000   

           DiffWalk  
count  70692.000000  
mean       0.252730  
std        0.434581  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

In [160]:
standardScaler = preprocessing.StandardScaler()
columns_to_scale= ['BMI','GenHlth','MentHlth','PhysHlth']
data[columns_to_scale] = standardScaler.fit_transform(data[columns_to_scale])

In [161]:
x = data.drop(['Diabetes_binary'],axis = 1)
y = data['Diabetes_binary']



In [162]:
x_full,xtest_full,y_full,y_test_full = model_selection.train_test_split(x,y,test_size=0.1,random_state=32,shuffle=True)
#reducing dataset size to accomodate for processing.
x_train,x_test,y_train,y_test=model_selection.train_test_split(xtest_full,y_test_full,test_size=0.33,random_state=32)

#Converting pd dataframes to numpy arrays
x_train,x_test,y_train,y_test=x_train.to_numpy(),x_test.to_numpy(),y_train.to_numpy(),y_test.to_numpy()

m = y_train.shape
y_train = y_train.reshape(m[0],1)
y_test = y_test.reshape(y_test.shape[0],1)
x_train,x_test,y_train,y_test= x_train.T,x_test.T,y_train.T,y_test.T

print(x_train.shape)
print(y_train.shape)
print(x_test.shape)

(17, 4736)
(1, 4736)
(17, 2334)


Building the logistic regression model

In [163]:
def initialise(dim):
        '''Initialise the parameter values and bias values'''
        w = np.zeros((dim,1))
        b = 0
        print("w",w.shape)
        return w,b
        

def sigmoid(z):
    '''Compute sigmoid function'''
    sig = 1/(1+np.exp(-z))
    
    return sig

def cost_fn(w,b,x_train,y_train):
    '''Written cost function for both parameters and bias'''
    m = x_train.shape[1]
   
    
    A = sigmoid(np.dot(w.T,x_train)+ b)
   
    cst_fn = (-1)*(np.sum((np.log(A.T))*y_train+ (np.log(1-A.T))*(1-y_train)))/m
    
    
    

    dw = np.dot(x_train,(A-y_train).T)/m
    db = np.sum(A-y_train)/m
  

    gradient = {"dw":dw, "db": db}
   
    return cst_fn, gradient

    




Implementing Gradient descent

In [164]:
def gradient_descent(w,b,x_train,y_train,epochs,learning_rate):
    
    for i in range(epochs):
        costs = []
        cost,gradient = cost_fn(w,b,x_train,y_train)

        dw = gradient["dw"]
        db= gradient["db"]

        w = w- (learning_rate*dw)
        b= b - (learning_rate*db)

        if i%50==0:
            costs.append(cost)
    parameters = {"w":w,"b":b}
    gradient = {"dw":dw,"db":db}
    

    return  parameters, gradient, costs
    




Writing the predicting function 

In [165]:
def predict(w,b,x_test):
    m = x_test.shape[1]
    y_prediction = np.zeros((1,m))
    w = w.reshape(x_test.shape[0],1)
    
    A = sigmoid(np.dot(w.T,x_test)+b)

    for i in range(A.shape[1]):
        if A[0,i]>=0.5:
            y_prediction[0,i]=1
        else:
            y_prediction[0,i]=0
    
    return y_prediction

Combining preceeding functions to make the logistic regression model 

In [166]:
def logistic_regression(x_train, y_train,x_test,y_test,epochs=100,learning_rate=3):
    w,b = initialise(x_train.shape[0])

    parameters,gradient,costs= gradient_descent(w,b,x_train,y_train,epochs,learning_rate)
    w = parameters["w"]
    b = parameters["b"]
    
    y_pred_train = predict(w,b,x_train)
    y_pred_test = predict(w,b,x_test)

    accuracy_train= (100 - np.mean(np.abs(y_pred_train-y_train)*100))
    accuracy_test =(100 - np.mean(np.abs(y_pred_test-y_test)*100))

    print("Accuracy on training data", accuracy_train)
    print("Accuracy on test data", accuracy_test)
    return parameters


Training and testing the data.
The following step may take a while 


In [169]:
model = logistic_regression(x_train, y_train,x_test,y_test,epochs=50,learning_rate=0.5)



w (17, 1)
Accuracy on training data 72.50844594594595
Accuracy on test data 72.3650385604113


The model has the following accuracy 
Accuracy on training data 72.50844594594595
Accuracy on test data 72.3650385604113

In [ ]:

#Writing the data onto a file 
f1 = open("model_parameters.txt","w")
f1.write(str(model))
f1.close()